docs/fn-2.0/README.md

```diff
@@ -0,0 +1,416 @@
+# Fn 2.0
+
+## NOTE: This is a proposal for updating Fn with the following goals:
+
+* Split HTTP routes from functions
+* Support CloudEvent messages as the native format throughout
+* Simplify everything
+* Modularize components
+* Namespacing
+
+## Removals
+
+* Default, HTTP and JSON formats will be removed
+* Cold functions will be removed
+* Async, queues are event sources now
+
+## Terminology
+
+Fn follows the terminology defined here: https://github.com/cloudevents/spec/blob/master/spec.md#notations-and-terminology
+
+As well as the following:
+
+### Event Sources
+
+Event sources are things that emit events and pass the event messages to Fn. Three common event sources are:
+
+* HTTP
+* Scheduler
+* Queue
+
+### Functions
+
+Functions are little programs that accept a defined input and output a defined output. They do not need to be aware of any other parts
+of the system and can run standalone; from the command line, etc. These can be written in any language and are packaged up as a container image. 
+
+* Functions read from STDIN, write to STDOUT and log to STDERR
+* Functions are packaged in container images
+* Input is a stream of CloudEvent messages in JSON format
+* Output is a stream of CloudEvent messages in JSON format
+  * With some additions to deal with things like error handling and protocol specific responses (eg: function wants to return a specific http status code if it was fired from an HTTP event source.)
+
+#### Function Error Handling
+
+Errors will be returned as a CloudEvent with an “error” attribute, no “data” attribute, describing the error. 
+
+```
+{
+  "cloud-events-version": "0.1",
+  "event-id": "6480da1a-5028-4301-acc3-fbae628207b3",
+  "source": "http://example.com/repomanager",
+  "event-type": "http-request",
+  "event-type-version": "v1.5",
+  "event-time": "2018-04-01T23:12:34Z",
+  "content-type": "application/json",
+  “error”: {
```

Comment:
CloudEvents spec does not seem to have Error as part of it... Would be nice if we could A stick to the format or B Get errors added to the spec if the team in charge of it seems to agree that it would be important.

Reply:
That's true. If we want to return an error from the function we definitely need to use extensions.
At this moment the function is responsible for sending back complete CloudEvent object, which looks quite complex and unnecessary because the protocol dispatcher can do that for each function's call.
IMHO there is no strong requirement for a function to write back complete CloudEvent.

Reply:
@denismakogon I disagree the container should send entire cloud event back on stdout since it consumes an entire one. It should just be able to easily create them using the fdk so that alot of the heavy lifting is done by the fdk.

Reply:
Part of the spec is that you can add whatever you want to it. Recommended to be in extensions, but I think errors are important enough to be at the top level.

Reply:
re: errors, i plan to open an issue shortly here https://github.com/cloudevents/spec/issues to discuss this with them, this seems like something that is useful to add. but also, yeah, the spec mentions you can add fields outside of extensions, but hopefully we can get first class support soon enough and use that. stay posted.

Reply:
Ok :)